plugin/semisync/semisync_master.cc

```diff
@@ -554,10 +579,121 @@ bool ReplSemiSyncMaster::is_semi_sync_slave() {
   return val;
 }
 
-void ReplSemiSyncMaster::reportReplyBinlog(const char *log_file_name,
-                                           my_off_t log_file_pos) {
+// This method was copied from get_slave_uuid() in rpl_master.cc
+std::string ReplSemiSyncMaster::get_slave_uuid(THD *thd) const {
+  std::string result;
+
+  /* Protects thd->user_vars. */
+  mysql_mutex_lock(&thd->LOCK_thd_data);
+
+  const auto it = thd->user_vars.find("slave_uuid");
+  if (it != thd->user_vars.end() && it->second->length() > 0) {
+    result = std::string(it->second->ptr(), it->second->length());
+  }
+
+  mysql_mutex_unlock(&thd->LOCK_thd_data);
+  return result;
+}
+
+bool ReplSemiSyncMaster::update_whitelist(std::string &wlist) {
+  std::lock_guard<std::mutex> guard(rpl_semi_sync_master_whitelist_set_lock);
+
+  // remove all spaces
+  wlist.erase(std::remove(wlist.begin(), wlist.end(), ' '), wlist.end());
+
+  whitelist_set_t local_whitelist_set = rpl_semi_sync_master_whitelist_set;
+
+  // case: add a single uuid to the whitelist (value starts with +)
+  if (wlist[0] == '+') {
+    const auto str = wlist.substr(1);
+    // case: +ANY specified, start with a clean slate
+    if (str == any_slave) {
+      local_whitelist_set.clear();
+    }
+    local_whitelist_set.insert(str);
+  }
+  // case: remove a single uuid to the whitelist (value starts with -)
+  else if (wlist[0] == '-') {
+    const auto str = wlist.substr(1);
+    // case: -ANY specified, start with a clean slate
+    if (str == any_slave) {
+      local_whitelist_set.clear();
+    } else {
+      local_whitelist_set.erase(str);
+    }
+  }
+  // case: full comma separated string is specified
+  else {
+    local_whitelist_set = split_into_set(wlist, ',');
+  }
+
+  // re-calculate wlist from the set
+  wlist = "";
+  for (const auto &uuid : local_whitelist_set) {
+    wlist.append(uuid);
+    wlist.append(",");
+  }
+  // remove last comma
+  if (!wlist.empty()) {
+    wlist.pop_back();
+  }
+
+  rpl_semi_sync_master_whitelist_set = local_whitelist_set;
+  // remove ANY from set
+  rpl_semi_sync_master_whitelist_set.erase(any_slave);
+  // NOTE: make sure to change the version only after the updating the whitelist
+  ++rpl_semi_sync_master_whitelist_ver;
+
+  LogErr(INFORMATION_LEVEL, ER_SEMISYNC_WHITELIST_UPDATED,
+         rpl_semi_sync_master_whitelist, wlist.c_str());
+
+  rpl_semi_sync_master_whitelist_is_any = !strcmp(wlist.c_str(), any_slave);
```

Comment:
I think technically, this cached value of the whitelist should be updated before the version variable is bumped up. Otherwise, it's possible the slave would skip the any_slave check in verify_against_whitelist(), see the new version number, check the list (which is "ANY"), but not find its UUID in there, and be denied access. Though this problem would be resolved on the next connect.

Reply:
Both `verify_against_whitelist` and `update_whitelist` are protected with 
```
std::lock_guard<std::mutex> guard(rpl_semi_sync_master_whitelist_set_lock)
```

Reply:
Sorry, I missed that the guard was moved outside the if check in verify_against_whitelist(). The original code used the version comparison as an optimization to avoid having to grab the lock on every verify_against_whitelist() call though.  I think it would be better to follow the original implementation for this.

Reply:
Using `current_thd->semisync_whitelist_ver` is not possible from the AckReceiver thread therefore I moved `whitelist_ver` to  a new `semisync_info_t` struct.

Reply:
Ah, okay. Hm... comparing UUID would be more expensive than server_id then for access to the map.

Looking at this code, can we get rid of rpl_semi_sync_master_whitelist_set_lock if update_whitelist() also called lock() to acquire the LOCK_binlog_? It already looks like all of the other calls to update/modify the white list already acquire this lock, so we could use it to protect the set.

I think the whitelist set could be part of AckReceiver. We could add a verified field to struct Slave that caches the check. Whenever the whitelist is changed, go through the list of m_slaves to check the UUIDs. Whenever AckReceiver::add_slave() is called, compare it with the whitelist set. If the whitelist set is only accessed with the m_lock mutex in AckReceiver, we can remove the whitelist_set mutex.

This might be a simpler solution?